### PR TITLE
Amp action macro cleanup

### PR DESCRIPTION
**Pull request by alabiaga**
- do not allow a macro to recursively call itself.
- add examples page.

### CODE REVIEW

**Comment by alabiaga** [comment]
@choumx Hey Will note that you referred to the amp-bind-macro and its check for ensuring it refers to macros that were predefined before it. For the amp-action macro  this isn't an issue as when we hit the trigger code path in the action macro, it is already guaranteed to be defined. The existence of the target/amp-component that it will invoke is also getting checked in the actions service. 
https://github.com/ampproject/amphtml/blob/master/src/service/action-impl.js#L524

File: extensions/amp-action-macro/0.1/amp-action-macro.js [hunk 1]
```diff
@@ -87,6 +88,18 @@ export class AmpActionMacro extends AMP.BaseElement {
...
+  /**
+   * Checks if the action is triggering itself.
+   * @param {!../../../src/service/action-impl.ActionInvocation} invocation
+   * @return {boolean}
+   * @private
+   */
+  isRecursiveInvocation_(invocation) {
+    return invocation.source.getAttribute('id')
+        === this.element.getAttribute('id')
+        && invocation.source.tagName === 'AMP-ACTION-MACRO';
```
Comment (choumx) [inline extensions/amp-action-macro/0.1/amp-action-macro.js]: We need to worry about cycles generally. I.e. not just `x -> x` but `x -> y -> x`.

**Comment by alabiaga** [comment]
Gotcha but why not just not let the JS engine max call stack count handle
this? To me this is more of an error on the developers side.

File: extensions/amp-action-macro/0.1/amp-action-macro.js [hunk 2]
```diff
@@ -87,6 +100,28 @@ export class AmpActionMacro extends AMP.BaseElement {
...
+
+  /**
+   * Checks if the action is triggering itself.
+   * @param {!../../../src/service/action-impl.ActionInvocation} invocation
+   * @return {boolean}
+   * @private
+   */
+  isRecursiveInvocation_(invocation) {
+    return invocation.caller.getAttribute('id')
+        === this.element.getAttribute('id');
```
Comment (choumx) [inline extensions/amp-action-macro/0.1/amp-action-macro.js]: The other check should cover this case.

File: extensions/amp-action-macro/0.1/amp-action-macro.js [hunk 3]
```diff
@@ -87,6 +100,28 @@ export class AmpActionMacro extends AMP.BaseElement {
...
+
+  /**
+   * Checks if invoking action that was defined after it. This constraint is to
+   * prevent possible recursive calls.
+   * @param {number} invocationRegistrationId
+   * @return {boolean}
+   * @private
+   */
+  isInvalidMacroReference_(invocationRegistrationId) {
+    return invocationRegistrationId <= this.element['registrationId'];
```
Comment (choumx) [inline extensions/amp-action-macro/0.1/amp-action-macro.js]: Can we use this instead? https://developer.mozilla.org/en-US/docs/Web/API/Node/compareDocumentPosition

File: extensions/amp-action-macro/0.1/amp-action-macro.js [hunk 4]
```diff
@@ -75,9 +81,16 @@ export class AmpActionMacro extends AMP.BaseElement {
             arg, this.element);
       }
     }
+    if (invocation.caller.tagName === 'AMP-ACTION-MACRO') {
```
Comment (choumx) [inline extensions/amp-action-macro/0.1/amp-action-macro.js]: `invocation.caller.tagName.toLowerCase() === TAG`

File: extensions/amp-action-macro/0.1/amp-action-macro.js [hunk 5]
```diff
@@ -87,6 +100,28 @@ export class AmpActionMacro extends AMP.BaseElement {
...
+  }
+
+  /**
+   * Checks if invoking action that was defined after it. This constraint is to
+   * prevent possible recursive calls.
+   * @param {number} invocationRegistrationId
+   * @return {boolean}
+   * @private
+   */
+  isInvalidMacroReference_(invocationRegistrationId) {
```
Comment (choumx) [inline extensions/amp-action-macro/0.1/amp-action-macro.js]: Nit: Single caller of a simple function -- can be inlined.

File: extensions/amp-action-macro/0.1/amp-action-macro.js (hunk 2 above)
Comment (alabiaga) [reply extensions/amp-action-macro/0.1/amp-action-macro.js]: done.

File: extensions/amp-action-macro/0.1/amp-action-macro.js (hunk 3 above)
Comment (alabiaga) [reply extensions/amp-action-macro/0.1/amp-action-macro.js]: This is new to me. Fancy. done.

**Comment by alabiaga** [comment]
@choumx friendly ping

File: extensions/amp-action-macro/0.1/amp-action-macro.js [hunk 6]
```diff
@@ -75,9 +75,13 @@ export class AmpActionMacro extends AMP.BaseElement {
             arg, this.element);
       }
     }
+    if (invocation.caller.tagName.toLowerCase() === TAG) {
+      userAssert(this.isValidMacroReference_(
+          invocation.caller),
```
Comment (choumx) [inline extensions/amp-action-macro/0.1/amp-action-macro.js]: Nit: Move to previous line.

File: extensions/amp-action-macro/0.1/amp-action-macro.js [hunk 7]
```diff
@@ -75,9 +75,13 @@ export class AmpActionMacro extends AMP.BaseElement {
...
+      'Action macro with ID "%s" cannot reference itself or macros defined '
+          + 'after it', this.element.getAttribute('id'));
```
Comment (choumx) [inline extensions/amp-action-macro/0.1/amp-action-macro.js]: "amp-action-macro#%s may only reference macros that precede it."

**Comment by alabiaga** [comment]
thanks!

**Pull request merged**